<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 5 Looking ahead | Portfolio</title>
<meta name="author" content="Lunar Eclipse">
<meta name="description" content="5.1 Single Cell NGS To demonstrate my ability to learn new skills and continue to develop myself I will try to learn as much as I can in four days about single cell next generation sequencing. I...">
<meta name="generator" content="bookdown 0.27 with bs4_book()">
<meta property="og:title" content="Chapter 5 Looking ahead | Portfolio">
<meta property="og:type" content="book">
<meta property="og:description" content="5.1 Single Cell NGS To demonstrate my ability to learn new skills and continue to develop myself I will try to learn as much as I can in four days about single cell next generation sequencing. I...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 5 Looking ahead | Portfolio">
<meta name="twitter:description" content="5.1 Single Cell NGS To demonstrate my ability to learn new skills and continue to develop myself I will try to learn as much as I can in four days about single cell next generation sequencing. I...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.3.1/transition.js"></script><script src="libs/bs3compat-0.3.1/tabs.js"></script><script src="libs/bs3compat-0.3.1/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<style type="text/css">
    /* Used with Pandoc 2.11+ new --citeproc when CSL is used */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
        }
    .hanging div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }
  </style>
<link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">Portfolio</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"><span class="header-section-number">1</span> Welcome!</a></li>
<li><a class="" href="example-analysis-c.-elegans-experiment.html"><span class="header-section-number">2</span> Example analysis C. Elegans experiment</a></li>
<li><a class="" href="open-peer-review.html"><span class="header-section-number">3</span> Open Peer Review</a></li>
<li><a class="" href="guerrilla-analytics.html"><span class="header-section-number">4</span> Guerrilla Analytics</a></li>
<li><a class="active" href="looking-ahead.html"><span class="header-section-number">5</span> Looking ahead</a></li>
<li><a class="" href="projects.html"><span class="header-section-number">6</span> Projects</a></li>
<li><a class="" href="resume.html"><span class="header-section-number">7</span> Resume</a></li>
<li><a class="" href="relational-databases.html"><span class="header-section-number">8</span> Relational Databases</a></li>
<li><a class="" href="parameters.html"><span class="header-section-number">9</span> Parameters</a></li>
<li><a class="" href="normalizeme.html"><span class="header-section-number">10</span> normalizeme</a></li>
<li><a class="" href="references.html"><span class="header-section-number">11</span> References</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/LunarEclipse112/lunareclipse.github.io">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="looking-ahead" class="section level1" number="5">
<h1>
<span class="header-section-number">5</span> Looking ahead<a class="anchor" aria-label="anchor" href="#looking-ahead"><i class="fas fa-link"></i></a>
</h1>
<div id="single-cell-ngs" class="section level2" number="5.1">
<h2>
<span class="header-section-number">5.1</span> Single Cell NGS<a class="anchor" aria-label="anchor" href="#single-cell-ngs"><i class="fas fa-link"></i></a>
</h2>
<p>To demonstrate my ability to learn new skills and continue to develop myself I will try to learn as much as I can in four days about single cell next generation sequencing.</p>
<p>I chose this skill because NGS is something that I find very interesting and NGS from a single cell sounds awesome.</p>
<div id="plan" class="section level3" number="5.1.1">
<h3>
<span class="header-section-number">5.1.1</span> Plan<a class="anchor" aria-label="anchor" href="#plan"><i class="fas fa-link"></i></a>
</h3>
<p>First a small overview of how I want to structure my 4 days to make sure I best make use of the time. By the end of it I can hopefully run a analysis, or part of one that isn’t too complicated. To this end I will need to read up about it and search for usable data used in research to try and mimic what they have done. Hopefully in the future this will allow me to better develop the skills needed to run my own analysis on Single Cell NGS data.</p>
<div class="inline-table"><table class="table table-sm">
<colgroup>
<col width="8%">
<col width="83%">
<col width="8%">
</colgroup>
<thead><tr class="header">
<th align="left">Day Number</th>
<th align="left">Activity to do</th>
<th align="left">Completion</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Day 1</td>
<td align="left">Reading about single cell NGS and looking for research articles where this has been used for analysis</td>
<td align="left">DONE</td>
</tr>
<tr class="even">
<td align="left">Day 2, 3</td>
<td align="left">Trying to get everything to work to mimic an analysis found in the article</td>
<td align="left">DONE</td>
</tr>
<tr class="odd">
<td align="left">Day 4</td>
<td align="left">Continue with the reproduction of the article analysis untill done</td>
<td align="left">DONE</td>
</tr>
<tr class="even">
<td align="left">Day 5</td>
<td align="left">Write a report on what I have learned and things I need to learn more about, a what’s next secion let’s say</td>
<td align="left">DONE</td>
</tr>
</tbody>
</table></div>
</div>
<div id="next-generation-sequencing" class="section level3" number="5.1.2">
<h3>
<span class="header-section-number">5.1.2</span> Next generation sequencing<a class="anchor" aria-label="anchor" href="#next-generation-sequencing"><i class="fas fa-link"></i></a>
</h3>
<p>Next generation sequencing (NGS) or massively parallel sequencing is a technology which allows for fast and precise sequencing of DNA or cDNA. This is done by parallel sequencing of a very high number of DNA (or cDNA) fragments and provides high depth and much insight into genomes of organisms or can be used to explore gen expression in a single organism. With bioinformatic analysis the data acquired can be put together to form a complete genome or give insight in to cellular activity of an organism (<span class="citation">Behjati and Tarpey (2013)</span>). The technology has brought many different advances in science and is now being looked at to be implemented in the diagnostics field as a noninvasive tool to diagnose infectious diseases in immunocompromised hosts (<span class="citation">Jose F. Camargo 1, Asim A. Ahmed 2, Martin S. Lindner2, Michele I. Morris1, Shweta Anjan1, Anthony D. Anderson 3, Clara E. Prado4, Sudeb C. Dalai2, and Octavio V. Martinez4, Krishna V. Komanduri (n.d.)</span>).</p>
<p>Although the technology is very impressive and very useful for a wide range of applications (<span class="citation">Behjati and Tarpey (2013)</span>) it has a significant drawback. While a picture can be formed when looking at RNA expression in tissue or type specific cells it’s always clouded due to the interference of different cells in the same “sample”. By design NGS utilizes a very large amount of cDNA fragments that need to be analysed <em>in silico</em>
to align the fragments to a reference genome and gain information about the genes expressed. The problem arises when one wants to more accurately look at how much a certain gene was expressed. Due to differences in cellular expression between cells the accuracy of determination of gene expression will always be limited. Two cells of the same type might be producing different quantities of RNA that is used measured in the NGS analysis (as cDNA) but these differences can’t be retraced because each run utilizes cDNA fragments from multiple cells. As a solution to this problem (among others) is to run a differential gene expression analysis (DGE). The DGA analysis provides information about the statistical difference in expressed genes based on a probability value. It is a reliable method to find the <em>Log2 fold changes</em> which quite accurately calculate an answer to the how much question of the problem. Another solution is single cell next generation sequencing which also provides answers to this problem.</p>
</div>
<div id="single-cell-next-generation-sequencing" class="section level3" number="5.1.3">
<h3>
<span class="header-section-number">5.1.3</span> Single cell next generation sequencing<a class="anchor" aria-label="anchor" href="#single-cell-next-generation-sequencing"><i class="fas fa-link"></i></a>
</h3>
<p>Single cell transcriptomics is used to characterize cell states (<span class="citation">Stuart et al. (2019)</span>). Ranging from gene expression and spatial transcriptomics to chromatine accesibility and immunophenotyping scRNA-seq technology has a broad spectrum of applications, each with their own strengths and weaknesses. Most measure only a very specific aspect of cellular identity posing a clear need to use the information acquired from one analysis to strengthen the conclusion of another.</p>
</div>
<div id="scatac-seq" class="section level3" number="5.1.4">
<h3>
<span class="header-section-number">5.1.4</span> scATAC-seq<a class="anchor" aria-label="anchor" href="#scatac-seq"><i class="fas fa-link"></i></a>
</h3>
<p>Single cell NGS has been used in research to better understand the developmental programs encoded in the linear genome sequence (<span class="citation">Fang et al. (2021)</span>). Single cell NGS can also be used to identify the <em>cis</em>-elements in organisms and specific cell types. Cell type is established due to regulation of spatiotemporal gene expression programs (<span class="citation">The ENCODE Project Consortium (2012)</span>). These interact with sequence specific transcription factors and <em>cis-</em>regulatory sequences in a cell-type specific manner. Due to sensitivity to nucleases and transposases when active there have been different methods developed to find de <em>cis</em>-elements in organisms and specific cell types. Chromatine accesability methods are for example scTHIS-seq (single-cell transposome hypersensitive site sequencing) or scATAC-seq. In scATA-seq or single-cell ATAC-seq analysis allows for profiling of chromatine accessibility from hundreds of thousands cells in a single experiment (<span class="citation">Fang et al. (2021)</span>).</p>
<p>Current methods often require the use of linear dimensionality reduction on a cell matrix of thousands of dimensions. Scaling this to millions of cells is a huge challenge. In addition scATAC-seq datasets do not posses the same level of sensitivity as scRNA-seq. To overcome the limitations a single nucleus analysis pipeline for <a href="https://github.com/r3fang/SnapATAC/blob/master/README.md">ATAC-seq-SnapATAC</a> has been proposed which I will be trying out here.</p>
</div>
<div id="snapatac" class="section level3" number="5.1.5">
<h3>
<span class="header-section-number">5.1.5</span> SnapATAC<a class="anchor" aria-label="anchor" href="#snapatac"><i class="fas fa-link"></i></a>
</h3>
<p>As posed before I will now be trying out the package SnapATAC on a provided dataset 10X (by the same author). The package runs in python so this will take more time than the expected 4 days but that’s not a very big deal.</p>
</div>
<div id="python-in-r" class="section level3" number="5.1.6">
<h3>
<span class="header-section-number">5.1.6</span> Python in R<a class="anchor" aria-label="anchor" href="#python-in-r"><i class="fas fa-link"></i></a>
</h3>
<p>To use python code chunks in Rmarkdown I downloaded the reticulate package. This chunk below is running python.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb5-1"><a href="looking-ahead.html#cb5-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-2"><a href="looking-ahead.html#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Hello World"</span>)</span>
<span id="cb5-3"><a href="looking-ahead.html#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Hello World</span></span></code></pre></div>
<p>Now to download the SnapATAC2 package we will need pip or maybe Anaconda. I am not yet sure what is the better option so I am trying to get Anaconda in my environment so I can maybe download the package and run it trough Rstudio. I am also trying to get it to install with pip, which should be easier I think.
Sadly using pip has not worked. I can get pip to install packages just fine but SnapAtac2 doesn’t want to install. I checked the issues on the github page of the package and found that the rust and anndate package would be able to help so I downloaded them both but still nothing. I also read that snapatac2 hadn’t been tested for windows but that was an older message so maybe that’s not the issue. Lets proceed as if it is supposed to work and see if we can get it to do so.
With pip failing we are left with Anaconda. Let’s try do download a simple package first to see if it works properly. To check if anaconda works we installed the numpy package. The download was successful but downloading snapatac2 did not. It’s time to look for a different package, I would prefer to find one for R, I am more comfortable working with R packages and I have limited time.</p>
</div>
<div id="cellwalkr" class="section level3" number="5.1.7">
<h3>
<span class="header-section-number">5.1.7</span> CellWalkR<a class="anchor" aria-label="anchor" href="#cellwalkr"><i class="fas fa-link"></i></a>
</h3>
<p>The next package I found is CellWalkR. <a href="https://github.com/PFPrzytycki/CellWalkR/blob/master/examples/CellWalkR_Vignette.md">CellWalkR</a> is a package to combine scATAC-seq data with labels and other epigenetic data. We will be using the data linked in the vignette which has been pre processed with ArchR. We will also install ArchR for future use.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Package requires devtools and BiocManager</span>
<span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
<span class="co"># Then install Archr</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org/reference/remote-reexports.html">install_github</a></span><span class="op">(</span><span class="st">"GreenleafLab/ArchR"</span>, ref<span class="op">=</span><span class="st">"master"</span>, repos <span class="op">=</span> <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/repositories.html">repositories</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
<span class="co"># Lastly install the dependencies</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">ArchR</span><span class="op">)</span>
<span class="fu">ArchR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ArchR/man/installExtraPackages.html">installExtraPackages</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>The data linked is from <a href="https://www.nature.com/articles/s41586-021-03209-8">this</a> article (<span class="citation">Ziffra et al. (2021)</span>). I won’t go into depth about the article just now. For now we are just going to collect the data and follow the vignette tutorial.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">### Load in the Data ###########################################################</span>
  <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">ArchR</span><span class="op">)</span>
  <span class="fu">downloadSampleData</span><span class="op">(</span><span class="st">"vignette_data"</span><span class="op">)</span>
<span class="co">#&gt; [1] "vignette_data"</span>
<span class="co">#&gt; [1] "vignette_data"</span>
  <span class="va">ATACMat</span> <span class="op">&lt;-</span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/externalFormats.html">readMM</a></span><span class="op">(</span><span class="st">"vignette_data/SamplePeakMat.mtx"</span><span class="op">)</span>
  <span class="va">peaks</span> <span class="op">&lt;-</span> <span class="fu">as</span><span class="op">(</span><span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html">fread</a></span><span class="op">(</span><span class="st">"vignette_data/SamplePeaks.txt"</span>, header <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">$</span><span class="va">V1</span>, <span class="st">"GRanges"</span><span class="op">)</span>

<span class="co">### Define Label Nodes #########################################################</span>

  <span class="va">pathToLabels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"SampleMarkers1.txt"</span>, package <span class="op">=</span> <span class="st">"CellWalkR"</span><span class="op">)</span>
  <span class="va">labelGenes</span> <span class="op">&lt;-</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html">fread</a></span><span class="op">(</span><span class="va">pathToLabels</span><span class="op">)</span>

<span class="co"># Check the labeling data. It has to have at least two columns</span>
  <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">head</a></span><span class="op">(</span><span class="va">labelGenes</span><span class="op">)</span>
<span class="co">#&gt;    entrez  cluster   avg_diff</span>
<span class="co">#&gt; 1:  10299 RG-early -1.2297890</span>
<span class="co">#&gt; 2:   6167 RG-early  0.2546596</span>
<span class="co">#&gt; 3:  11168 RG-early  0.2570446</span>
<span class="co">#&gt; 4:   8760 RG-early -0.2578798</span>
<span class="co">#&gt; 5:   8503 RG-early  0.2613031</span>
<span class="co">#&gt; 6:  10208 RG-early  0.2618003</span>
  
<span class="co">### Building a Network #########################################################</span>
  
<span class="co"># Next we compute cell-to-cell similarity in order to build edges in the cell-to-cell proportion of the graph</span>
  
  <span class="va">cellEdges</span> <span class="op">&lt;-</span> <span class="fu">computeCellSim</span><span class="op">(</span><span class="va">ATACMat</span><span class="op">)</span>
  
<span class="co"># Check cellEdges </span>
  <span class="va">cellEdges</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>
<span class="co">#&gt; 5 x 5 sparse Matrix of class "dgCMatrix"</span>
<span class="co">#&gt;                                                            </span>
<span class="co">#&gt; [1,] 1.00000000 0.11338151 0.15001705 0.09244314 0.09813385</span>
<span class="co">#&gt; [2,] 0.11338151 1.00000000 0.14950372 0.08130564 0.09035017</span>
<span class="co">#&gt; [3,] 0.15001705 0.14950372 1.00000000 0.08960442 0.11350499</span>
<span class="co">#&gt; [4,] 0.09244314 0.08130564 0.08960442 1.00000000 0.06237177</span>
<span class="co">#&gt; [5,] 0.09813385 0.09035017 0.11350499 0.06237177 1.00000000</span>
  
<span class="co">### Computing Label-Cell Edges #################################################</span>
  
<span class="co"># Now we need to define which genomic regions correspond to which genes. We’ll use hg38 with Entrez identifiers in this example, using full gene bodies</span>
  <span class="va">regions</span> <span class="op">&lt;-</span> <span class="fu">getRegions</span><span class="op">(</span>geneBody <span class="op">=</span> <span class="cn">TRUE</span>, genome <span class="op">=</span> <span class="st">"hg38"</span>, names <span class="op">=</span> <span class="st">"Entrez"</span><span class="op">)</span>
<span class="co"># Check it</span>
  <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">head</a></span><span class="op">(</span><span class="va">regions</span><span class="op">)</span>
<span class="co">#&gt; GRanges object with 6 ranges and 1 metadata column:</span>
<span class="co">#&gt;             seqnames            ranges strand |     gene_id</span>
<span class="co">#&gt;                &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;</span>
<span class="co">#&gt;           1    chr19 58362552-58364751      - |           1</span>
<span class="co">#&gt;          10     chr8 18389282-18391481      + |          10</span>
<span class="co">#&gt;         100    chr20 44652034-44654233      - |         100</span>
<span class="co">#&gt;        1000    chr18 28176931-28179130      - |        1000</span>
<span class="co">#&gt;   100009613    chr11 70075234-70077433      - |   100009613</span>
<span class="co">#&gt;   100009667    chr10 68010663-68012862      - |   100009667</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 595 sequences (1 circular) from hg38 genome</span>
<span class="co"># Now we map between this data and the peaks</span>
  <span class="va">ATACGenePeak</span> <span class="op">&lt;-</span> <span class="fu">mapPeaksToGenes</span><span class="op">(</span><span class="va">labelGenes</span>, <span class="va">ATACMat</span>, <span class="va">peaks</span>, <span class="va">regions</span><span class="op">)</span>
  <span class="va">labelEdges</span> <span class="op">&lt;-</span> <span class="fu">computeLabelEdges</span><span class="op">(</span><span class="va">labelGenes</span>, <span class="va">ATACMat</span>, <span class="va">ATACGenePeak</span><span class="op">)</span>
<span class="co"># Check it</span>
  <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">head</a></span><span class="op">(</span><span class="va">labelEdges</span><span class="op">)</span>
<span class="co">#&gt;      RG-early        oRG        tRG         vRG    RG-div1</span>
<span class="co">#&gt; [1,]        0 0.04201235 0.04275655 0.019920365 0.05198877</span>
<span class="co">#&gt; [2,]        0 0.03108851 0.02250872 0.008458454 0.03121237</span>
<span class="co">#&gt; [3,]        0 0.02902324 0.02383554 0.008164892 0.03279098</span>
<span class="co">#&gt; [4,]        0 0.03035606 0.01968420 0.009663494 0.03510777</span>
<span class="co">#&gt; [5,]        0 0.02334971 0.02856838 0.005171970 0.03755360</span>
<span class="co">#&gt; [6,]        0 0.03310487 0.02331299 0.008916823 0.03410891</span>
<span class="co">#&gt;         RG-div2</span>
<span class="co">#&gt; [1,] 0.04404968</span>
<span class="co">#&gt; [2,] 0.02986432</span>
<span class="co">#&gt; [3,] 0.03135008</span>
<span class="co">#&gt; [4,] 0.02916076</span>
<span class="co">#&gt; [5,] 0.03137661</span>
<span class="co">#&gt; [6,] 0.03400801</span>
  
<span class="co">### Tuning Label Edges #########################################################</span>
  
<span class="co"># We now have cell-to-cell edges and label-to-cell edges, we don’t know how to correctly weight the two relative to each other. The tuneEdgeWeights method will run CellWalker across a range of possible parameters and compute cell homogeneity for each</span>
  <span class="va">labelEdgesList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">labelEdges</span><span class="op">)</span>
  
  <span class="va">edgeWeights</span> <span class="op">&lt;-</span> <span class="fu">tuneEdgeWeights</span><span class="op">(</span><span class="va">cellEdges</span>, 
                              <span class="va">labelEdgesList</span>, 
                              labelEdgeOpts <span class="op">=</span> <span class="fl">10</span><span class="op">^</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">7</span>,<span class="fl">1</span><span class="op">)</span>, 
                              sampleDepth <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span>
<span class="co"># Check it</span>
  <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">head</a></span><span class="op">(</span><span class="va">edgeWeights</span><span class="op">[</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/setorder.html">order</a></span><span class="op">(</span><span class="va">edgeWeights</span><span class="op">$</span><span class="va">cellHomogeneity</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,<span class="op">]</span><span class="op">)</span>
<span class="co">#&gt;    Var1 cellHomogeneity</span>
<span class="co">#&gt; 4 1e+04     -0.02789777</span>
<span class="co">#&gt; 5 1e+05     -0.03954104</span>
<span class="co">#&gt; 7 1e+07     -0.04022707</span>
<span class="co">#&gt; 6 1e+06     -0.05134244</span>
<span class="co">#&gt; 3 1e+03     -0.06223973</span>
<span class="co">#&gt; 2 1e+02     -0.23520271</span>

<span class="co">### Making a cellWalk Object ###################################################</span>

<span class="co"># Now we generate a cellWalk object with the above tuned edge weight parameter</span>
  <span class="va">cellWalk</span> <span class="op">&lt;-</span> <span class="fu">walkCells</span><span class="op">(</span><span class="va">cellEdges</span>, 
                     <span class="va">labelEdgesList</span>, 
                     labelEdgeWeights <span class="op">=</span> <span class="fl">1e+07</span><span class="op">)</span>
  
<span class="co">### Adding Filters #############################################################</span>

<span class="co"># We may have some bulk epigenetic data that can help filter down which peaks are relevant to our analysis. We can tune weights on each filter to determine how significant it is to our data. For our example we have H3K4me3 data which indicates active promoters. Thus we apply this filter permissively (setting filterOut=FALSE) and at the whole gene level (filterGene=TRUE) rather than just to overlapping peaks. We can make a new cellWalk object using this filter:</span>
  
  <span class="co"># labelEdges &lt;- computeLabelEdges(labelGenes, </span>
  <span class="co">#                              ATACMat, </span>
  <span class="co">#                              ATACGenePeak,</span>
  <span class="co">#                              filters = filters, </span>
  <span class="co">#                              filterWeights = c(1),</span>
  <span class="co">#                              filterOut = c(FALSE),</span>
  <span class="co">#                              filterGene = c(TRUE),</span>
  <span class="co">#                              regions = regions)</span>
  <span class="co"># labelEdgesList &lt;- list(labelEdges)</span>
  <span class="co"># cellWalk &lt;- walkCells(cellEdges, </span>
  <span class="co">#                    labelEdgesList, </span>
  <span class="co">#                    labelEdgeWeights = 1e+07)</span>

<span class="co">### Downstream Analysis ########################################################</span>

<span class="co"># Once we have a cellWalk object we can use it for downstream analysis</span>
  
<span class="co"># The label threshold determines the minimum influence score a cell must get to be considered labeled</span>
  <span class="va">cellWalk</span> <span class="op">&lt;-</span> <span class="fu">findUncertainLabels</span><span class="op">(</span><span class="va">cellWalk</span>, labelThreshold <span class="op">=</span> <span class="fl">0</span>, plot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="04-looking_ahead_files/figure-html/unnamed-chunk-5-1.png" width="672"></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co"># We can also directly examine label similarity by considering label-to-label influence</span>
  <span class="va">cellWalk</span> <span class="op">&lt;-</span> <span class="fu">clusterLabels</span><span class="op">(</span><span class="va">cellWalk</span>,  plot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="04-looking_ahead_files/figure-html/unnamed-chunk-5-2.png" width="672"></div>
<pre><code>#&gt; NULL
  
  cellWalk &lt;- plotCells(cellWalk, labelThreshold = 0, seed = 1)</code></pre>
<div class="inline-figure"><img src="04-looking_ahead_files/figure-html/unnamed-chunk-5-3.png" width="672"></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R">    
<span class="co"># It is also possible to plot how strongly a single label influences each cell in the embedding</span>
  <span class="va">cellWalk</span> <span class="op">&lt;-</span> <span class="fu">plotCells</span><span class="op">(</span><span class="va">cellWalk</span>, cellTypes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="st">"RG-div2"</span><span class="op">)</span>, seed <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="04-looking_ahead_files/figure-html/unnamed-chunk-5-4.png" width="672"></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R">  
<span class="co"># Furthermore, to analyze rare cell types, it can be helpful to only plot a subset of of all labels</span>
  <span class="va">cellWalk</span> <span class="op">&lt;-</span> <span class="fu">plotCells</span><span class="op">(</span><span class="va">cellWalk</span>, cellTypes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="st">"RG-early"</span>,<span class="st">"tRG"</span>,<span class="st">"vRG"</span><span class="op">)</span>, labelThreshold <span class="op">=</span> <span class="fl">0</span>, seed <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="04-looking_ahead_files/figure-html/unnamed-chunk-5-5.png" width="672"></div>
</div>
<div id="whats-next" class="section level3" number="5.1.8">
<h3>
<span class="header-section-number">5.1.8</span> What’s Next<a class="anchor" aria-label="anchor" href="#whats-next"><i class="fas fa-link"></i></a>
</h3>
<p>As my next steps I would like to keep trying to get SnapAtac to work and also try the Cellwalker package with another research to actually run an analysis. As of now my time is done but I will continue to work on this subject as I find it very Interesting but from now on it will have a page of it’s own in my portfolio.
I learned a great deal about what scRNA-seq is but I am still barely scratching the surface. Most of the things I come across are rather new and I need a lot of time to read up.</p>

</div>
</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="guerrilla-analytics.html"><span class="header-section-number">4</span> Guerrilla Analytics</a></div>
<div class="next"><a href="projects.html"><span class="header-section-number">6</span> Projects</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#looking-ahead"><span class="header-section-number">5</span> Looking ahead</a></li>
<li>
<a class="nav-link" href="#single-cell-ngs"><span class="header-section-number">5.1</span> Single Cell NGS</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#plan"><span class="header-section-number">5.1.1</span> Plan</a></li>
<li><a class="nav-link" href="#next-generation-sequencing"><span class="header-section-number">5.1.2</span> Next generation sequencing</a></li>
<li><a class="nav-link" href="#single-cell-next-generation-sequencing"><span class="header-section-number">5.1.3</span> Single cell next generation sequencing</a></li>
<li><a class="nav-link" href="#scatac-seq"><span class="header-section-number">5.1.4</span> scATAC-seq</a></li>
<li><a class="nav-link" href="#snapatac"><span class="header-section-number">5.1.5</span> SnapATAC</a></li>
<li><a class="nav-link" href="#python-in-r"><span class="header-section-number">5.1.6</span> Python in R</a></li>
<li><a class="nav-link" href="#cellwalkr"><span class="header-section-number">5.1.7</span> CellWalkR</a></li>
<li><a class="nav-link" href="#whats-next"><span class="header-section-number">5.1.8</span> What’s Next</a></li>
</ul>
</li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/LunarEclipse112/lunareclipse.github.io/blob/master/04-looking_ahead.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/LunarEclipse112/lunareclipse.github.io/edit/master/04-looking_ahead.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>Portfolio</strong>" was written by Lunar Eclipse. It was last built on 2023-02-07.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>
</body>
</html>
